components/core/src/ffi/ir_stream/decoding_methods.cpp

```diff
@@ -409,6 +423,71 @@ namespace ffi::ir_stream {
         return IRErrorCode_Success;
     }
 
+    template <typename encoded_variable_t>
+    static std::string decode_message (
+            string_view logtype,
+            vector <encoded_variable_t> encoded_vars,
+            vector <string_view> dictionary_vars
+    ) {
+        std::string message;
+        size_t encoded_vars_length = encoded_vars.size();
+        size_t dict_vars_length = dictionary_vars.size();
+        size_t next_static_text_begin_pos = 0;
+
+        size_t dictionary_vars_ix = 0;
+        size_t encoded_vars_ix = 0;
+        for (size_t cur_pos = 0; cur_pos < logtype.length(); ++cur_pos) {
+            auto c = logtype[cur_pos];
+            if (enum_to_underlying_type(VariablePlaceholder::Float) == c) {
+                message.append(logtype, next_static_text_begin_pos, cur_pos - next_static_text_begin_pos);
+                next_static_text_begin_pos = cur_pos + 1;
+                if (encoded_vars_ix >= encoded_vars_length) {
+                    throw EncodingException(ErrorCode_Corrupt, __FILENAME__, __LINE__,
+                                            cTooFewEncodedVarsErrorMessage);
+                }
+                message.append(decode_float_var(encoded_vars[encoded_vars_ix]));
+                ++encoded_vars_ix;
+            } else if (enum_to_underlying_type(VariablePlaceholder::Integer) == c) {
+                message.append(logtype, next_static_text_begin_pos, cur_pos - next_static_text_begin_pos);
+                next_static_text_begin_pos = cur_pos + 1;
+                if (encoded_vars_ix >= encoded_vars_length) {
+                    throw EncodingException(ErrorCode_Corrupt, __FILENAME__, __LINE__,
+                                            cTooFewEncodedVarsErrorMessage);
+                }
+                message.append(decode_integer_var(encoded_vars[encoded_vars_ix]));
+                ++encoded_vars_ix;
+            } else if (enum_to_underlying_type(VariablePlaceholder::Dictionary) == c) {
+                message.append(logtype, next_static_text_begin_pos, cur_pos - next_static_text_begin_pos);
+                next_static_text_begin_pos = cur_pos + 1;
+                if (dictionary_vars_ix >= dict_vars_length) {
+                    throw EncodingException(ErrorCode_Corrupt, __FILENAME__, __LINE__,
+                                            cTooFewDictionaryVarsErrorMessage);
+                }
+                message.append(dictionary_vars[dictionary_vars_ix]);
+                ++dictionary_vars_ix;
+            } else if (cVariablePlaceholderEscapeCharacter == c) {
+                if (cur_pos == logtype.length() - 1) {
```

Comment:
Perhaps a comment above this to say `Ensure the escape character is followed by a character that's being escaped`?